[jit] avoid pre-convert tensor to cpu in pickling

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#38898 [jit] avoid pre-convert tensor to cpu in pickling**

Pickling will pickle the tensor meta info, and its up to the jit
exporter or other upstream who use the pickler to decide how to write
the actual tensor data.

This PR make we call getWritableTensorData in upper level so that rpc
and TensorPipe can leverge it with only pickling tensor meta data without
converting the tensor from GPU to CPU.

Differential Revision: [D21879866](https://our.internmc.facebook.com/intern/diff/D21879866)

wanchaol committed:
[jit] avoid pre-convert tensor to cpu in pickling

Pickling will pickle the tensor meta info, and its up to the jit
exporter or other upstream who use the pickler to decide how to write
the actual tensor data.

This PR make we call getWritableTensorData in upper level so that rpc
and TensorPipe can leverge it with only pickling tensor data without
converting the tensor from GPU to CPU

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/utils.cpp:
```diff
@@ -272,12 +271,13 @@ std::string wireSerialize(
     pickler.pushIValue(cloneSparseTensors(tensors));
     pickler.stop();
     // tensorData is in function scope so that the data() pointers stay valid.
```
update this comment too?

lw replied:
I think what this comment tried to say was that we need to keep the WritableTensorData objects alive until lines 298-300, as that is the moment where we actually copy the data from the tensor to the buffer that will be sent over the network. The reason for this, I think, is because there could be temporary tensors that are only kept alive from WritableTensorData and would be deallocated if WritableTensorData is destroyed. I see two such types of tensors:
- the ones created by cloneSparseTensors in case we're only sending a small slice of a larger tensor;
- the ones created by getWritableTensorData in case the original tensor was a CUDA tensor.

Since the ProcessGroup agent doesn't support CUDA tensors I think the only real case where this could happen is the first one. And in that case I think this diff would introduce a bug. I'm a bit surprised there is no test that caught that.

lw replied:
The `Ent` instances in the `entries` vector just contain raw pointers to the underlying data of the tensor. That data is however owned and managed by an `at::Tensor` instance and as soon as it goes out of scope it deallocates the memory, meaning that the raw pointer in the `Ent` becomes invalid and may lead to a page fault when used. We thus need to keep the `at::Tensor` instances alive as well until we have copied their data out of them. For most tensors that will be automatic because they are kept alive by the caller of this function. However, there are also temporary `at::Tensor` instances created within this function, namely the ones created by `cloneSparseTensors`. Those need to be stored somewhere in order to keep them alive. This is what the comment is saying. If you feel the wording could be improved go ahead! :)

wanchaol replied:
gotcha! thanks for the detailed explaination, I tried locally, this indeed trigger a bug with the `cloneSparseTensors` case, and it's captured by https://github.com/pytorch/pytorch/blob/fe88806784e05af554247c8a29467adb6d84cc5b/test/cpp/rpc/test_wire_serialization.cpp#L39

But apparently this cpp test is not run in CI by default, will need to figure out how to enable this on CI.

mrshenli commented on torch/csrc/distributed/rpc/utils.cpp:
```diff
@@ -394,7 +394,7 @@ TensorPipeEntry tensorpipeSerialize(const Message& rpcMessage) {
     const auto& tensorDataVec = pickler.tensorData();
```
@wanchaol would I be correct that these tensors would preserve its device affinity instead of always live on CPU? And this is because we no longer call `getWriteableTensorData` when pushing tensors into `tensor_data_` in JIT pickler?

wanchaol replied:
Yes that's right, the tensors will still alive on whatever device they were (no copy to cpu operations). tensorPipe serialize can use the tensors get from the pickler and pass around the data ptrs across GPUs.

mrshenli reviewed (approved):
LGTM! Thanks for helping!!

wanchaol committed:
Update on "[jit] avoid pre-convert tensor to cpu in pickling"


Pickling will pickle the tensor meta info, and its up to the jit
exporter or other upstream who use the pickler to decide how to write
the actual tensor data.

This PR make we call getWritableTensorData in upper level so that rpc
and TensorPipe can leverge it with only pickling tensor meta data without
converting the tensor from GPU to CPU.

[ghstack-poisoned]

mrshenli commented:
cc @lw

lw reviewed (commented):
Thanks, this looks good! (except for that one comment)

I'm wondering whether we can add a test to make sure this works as intended. Especially to make sure that we can un-pickle this correctly after moving the tensor to a different device (which I think was one of our main concerns). So I'm thinking of a test like this (in pseudo-code):
```
pickle, [tensors] = jit::pickle(obj_in)
moved_tensors = [t.to("cuda:2") for t in tensors]
obj_out = jit::unpickle(pickle, moved_tensors)

// Check that tensors in obj_out reside on cuda:2
// Check that, apart from the tensor's devices, obj_in and obj_out are identical
```

wanchaol committed:
Update on "[jit] avoid pre-convert tensor to cpu in pickling"


Pickling will pickle the tensor meta info, and its up to the jit
exporter or other upstream who use the pickler to decide how to write
the actual tensor data.

This PR make we call getWritableTensorData in upper level so that rpc
and TensorPipe can leverge it with only pickling tensor meta data without
converting the tensor from GPU to CPU.

[ghstack-poisoned]

wanchaol commented:
@lw regarding on the test, the jit pickler will pickle the tensor device on CPU, and then unpickler will try to restore it on CPU according to the pickled device meta info, so I think converting it to CUDA afterwards will fail to load them again on GPU. Also unpickler needs a `read_record` func that returns a valid data ptr on the corresponding device, I think it probably more valid to add this test when  we do the follow up PR on tensorPipe serde to have the special read_record with the pickled device.

wanchaol committed:
Update on "[jit] avoid pre-convert tensor to cpu in pickling"


Pickling will pickle the tensor meta info, and its up to the jit
exporter or other upstream who use the pickler to decide how to write
the actual tensor data.

This PR make we call getWritableTensorData in upper level so that rpc
and TensorPipe can leverge it with only pickling tensor meta data without
converting the tensor from GPU to CPU.

[ghstack-poisoned]
